File: bindings/csharp/examples/CameraRecord.cs
// CameraRecord.cs — Mediaway C# quick start.
//
// The `Mediaway.Common`/`Mediaway.Device.Camera`/`Mediaway.Device.Audio`/
// `Mediaway.Pipeline` packages this targets are real — see bindings/csharp/src/
// and adr/0004-domain-feature-split.md (mirrored at the C# layer: Camera,
// Desktop, Audio, and Hotplug are now separate packages, not one
// `Mediaway.Device`). See ScreenRecord.cs for the Desktop (Screen) source
// counterpart of this same shape.
//
// Revision note (matching the real, shipped API): each package keeps its
// own root namespace. `IVideoCapture.Width`/`Height` are `uint`, matching
// the native ABI's `uint32_t` geometry. `Mediaway.Device.Camera.VideoFrame`
// and `Mediaway.Pipeline.VideoFrame` are two DISTINCT types, not one shared
// record: the Camera frame owns a disposable native buffer (Zero-Copy poll
// output — see Mediaway.Common.Interop.NativeOwnedMemoryManager), while the
// Pipeline frame is a plain, non-disposable borrowed-input value type. So
// RecordAsync below explicitly builds a new Pipeline frame from each
// captured Camera frame's fields and disposes the Camera frame afterward,
// instead of a single-type `with` clone. `EncodeSession.Finish()` returns an
// `IMemoryOwner<byte>` for the same reason (see EncodeToMp4.cs's own
// revision note).
//
// Scenario: capture the default webcam + default microphone, encode the
// video into H.264 through the same "auto video encoder -> encode session"
// building blocks as EncodeToMp4.cs, and write a fragmented MP4 to disk.
//
// Run:
//     dotnet run

using System.Buffers;
using Mediaway.Common;
using Mediaway.Device;
using Mediaway.Device.Audio;
using Mediaway.Device.Camera;
using Mediaway.Pipeline;
using DeviceVideoFrame = Mediaway.Device.Camera.VideoFrame;
using PipelineVideoFrame = Mediaway.Pipeline.VideoFrame;

const int Fps = 30;
const int Seconds = 3;

try
{
    await RunRecordingAsync();
}
catch (CaptureUnavailableException ex)
{
    // Thrown by Camera.Open below — no webcam is an environment problem for
    // this scenario, not a normal branch (unlike a missing microphone,
    // handled at its own call site via TryOpen below).
    Console.WriteLine($"CameraRecord: {ex.Message} — exiting.");
}
catch (EncoderUnavailableException ex)
{
    // Thrown by AutoVideoEncoder.Open below.
    Console.WriteLine($"CameraRecord: {ex.Message} — exiting.");
}

static async Task RunRecordingAsync()
{
    // ── 1. Open capture sources ─────────────────────────────────────────

    // Device 0 = default/first camera. The capture settles on its own
    // stream geometry (whatever the camera actually negotiated), read back
    // below via Width/Height. No camera -> nothing to record -> Open()
    // (throws), caught above.
    using var videoCapture = Camera.Open(deviceIndex: 0, frameRate: new Rational(1, Fps));

    // Recording without audio is a valid degraded mode — log and carry on
    // rather than exiting. `using` on a null IAudioCapture? is a documented
    // no-op.
    using var audioCapture = Microphone.TryOpen(sampleRateTimeBase: new Rational(1, 48_000), out var audioError);
    if (audioCapture is null)
    {
        Console.WriteLine($"CameraRecord: microphone unavailable ({audioError}) — continuing without audio.");
    }

    uint width = videoCapture.Width;
    uint height = videoCapture.Height;
    Console.WriteLine($"CameraRecord: {width}x{height} camera, mic {(audioCapture is null ? "unavailable" : "ready")}");

    // ── 2. Open the encoder + encode session (same building blocks as EncodeToMp4.cs) ──

    var config = VideoEncodeConfig.CreateDefault(VideoCodec.H264, width, height, new Rational(1, Fps)) with
    {
        BitrateBps = 4_000_000,
    };

    using var encoder = AutoVideoEncoder.Open(config);
    using var session = EncodeSession.Open(encoder);

    // ── 3. Stream captured frames straight into the encoder ─────────────

    using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(Seconds));
    try
    {
        await RecordAsync(videoCapture, audioCapture, session, cts.Token);
    }
    catch (OperationCanceledException)
    {
        // Expected: the recording duration elapsed.
    }

    // ── 4. Finish encoding and write the result ──────────────────────────

    using IMemoryOwner<byte> mp4Bytes = session.Finish();
    await using (var outFile = File.Create("out_camera.mp4"))
    {
        await outFile.WriteAsync(mp4Bytes.Memory);
    }

    Console.WriteLine($"CameraRecord: {width}x{height} -> out_camera.mp4 ({mp4Bytes.Memory.Length} bytes)");
}

// ── Record loop ──────────────────────────────────────────────────────────────
//
// Every parameter here is an interface type (or the session built on one) — this
// method has no idea which concrete OS backend is underneath, so it works
// identically whether `video`/`audio` are camera+mic capture, screen+mic
// capture, or test doubles. Identical shape to ScreenRecord.cs's RecordAsync —
// duplicated rather than shared, matching this binding set's convention that
// each example file stays self-contained and copy-pastable on its own.
static async Task RecordAsync(IVideoCapture video, IAudioCapture? audio, EncodeSession session, CancellationToken token)
{
    Task audioDrainTask = audio is null ? Task.CompletedTask : DrainAudioAsync(audio, token);

    long pts = 0;
    await foreach (DeviceVideoFrame frame in video.ReadFramesAsync(token))
    {
        using (frame) // Device.VideoFrame owns a native buffer — dispose after this frame is written.
        {
            session.WriteFrame(new PipelineVideoFrame
            {
                Pts = pts++,
                Duration = 1,
                Width = frame.Width,
                Height = frame.Height,
                PixelFormat = frame.PixelFormat,
                Data = frame.Data,
            });
        }
    }

    await audioDrainTask;

    // Not wired into an audio track yet — this loop only demonstrates that the
    // async frame stream composes the same way for an audio source.
    static async Task DrainAudioAsync(IAudioCapture audio, CancellationToken token)
    {
        await foreach (AudioFrame frame in audio.ReadFramesAsync(token))
        {
            frame.Dispose(); // AudioFrame also owns a native buffer.
        }
    }
}

File: bindings/nodejs/examples/camera-record.ts
/**
 * Camera + mic capture -> encode -> fragmented MP4 — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `@mediaway/device` or `@mediaway/encoder` npm package
 * exists yet. This file shows the target ergonomics for a future Node.js binding
 * over Mediaway's C ABI (napi-rs native addon under the hood, wrapped in
 * idiomatic Node/TypeScript: typed configs, `Error` subclasses, explicit
 * `close()`). This is the Node.js host reaching native code through the C ABI —
 * Node runs on servers/desktops and calls OS camera/mic capture APIs through the
 * native addon; it does not use browser capture APIs. See ../README.md and
 * docs/spec/c-ffi.md.
 *
 * Builds a camera-recording pipeline from the same building blocks as the
 * auto-encode example (config -> open auto encoder -> open encode session ->
 * writeFrame -> finish), plus a device-capture layer, glued together by the
 * same small, platform-agnostic `record()` function used for screen capture —
 * it only knows about the `VideoCapture`/`AudioCapture` interfaces, never the
 * concrete OS backend or source (camera vs screen) underneath.
 *
 * Run (once the real packages exist):
 *   npx tsx camera-record.ts
 */

import { writeFile } from "node:fs/promises";
import type { Rational } from "@mediaway/common";
import {
  type AudioCapture,
  type VideoCapture,
  openCameraCapture,
  openMicrophoneCapture,
} from "@mediaway/device";
import {
  AutoVideoEncodeConfig,
  EncodeSession,
  EncoderUnavailableError,
  openAutoEncoder,
  type VideoFrame,
} from "@mediaway/encoder";

const FPS = 30;
const DURATION_MS = 3_000;
const OUTPUT_PATH = "out_camera.mp4";

/** Build one grey NV12 frame: width*height Y bytes, then width*height/2 interleaved UV bytes. */
function makeGreyNv12Frame(width: number, height: number, pts: number): VideoFrame {
  const ySize = width * height;
  const uvSize = ySize / 2;
  const data = Buffer.alloc(ySize + uvSize, 0x80); // Y=128, U=128, V=128 everywhere

  return {
    pts,
    duration: 1,
    width,
    height,
    pixelFormat: "nv12",
    data,
  };
}

/**
 * Poll video + (optional) audio capture and feed the encode session until
 * `durationMs` elapses.
 *
 * This is the reusable "record loop": it only depends on the
 * `VideoCapture`/`AudioCapture` interfaces and an already-open `EncodeSession`,
 * so the exact same function works unchanged whether `video` came from a
 * camera or a screen capture backend.
 */
async function record(
  video: VideoCapture,
  audio: AudioCapture | undefined,
  session: EncodeSession,
  durationMs: number,
): Promise<void> {
  const deadline = Date.now() + durationMs;
  let frameIndex = 0;

  while (Date.now() < deadline) {
    // A real capture frame may reference GPU-resident memory (Zero-Copy handle);
    // this example only writes a synthetic placeholder into the encoder instead
    // of the real pixels, but still models the release step below.
    const videoFrame = video.pollFrame();
    if (videoFrame !== undefined) {
      try {
        await session.writeFrame(makeGreyNv12Frame(video.width, video.height, frameIndex));
        frameIndex++;
      } finally {
        // Release the GPU-resident buffer back to the OS before polling again.
        videoFrame.release();
      }
    }

    // Drain any polled audio frames — this example doesn't wire audio into the
    // MP4 yet, just keeps the capture queue from backing up.
    if (audio !== undefined) {
      while (audio.pollFrame() !== undefined) {
        // intentionally discarded
      }
    }

    if (videoFrame === undefined) {
      // Nothing new yet — yield briefly instead of busy-spinning.
      await new Promise<void>((resolve) => setTimeout(resolve, 1));
    }
  }
}

async function main(): Promise<void> {
  let video: VideoCapture;
  try {
    // 0 = default/first camera; capture settles on its own actual stream geometry.
    video = await openCameraCapture({ deviceIndex: 0, timeBase: { num: 1, den: FPS } satisfies Rational });
  } catch (err) {
    console.error(`camera-record: camera device unavailable (${(err as Error).message})`);
    return;
  }

  let audio: AudioCapture | undefined;
  try {
    audio = await openMicrophoneCapture({ timeBase: { num: 1, den: 48_000 } satisfies Rational });
  } catch (err) {
    console.warn(`camera-record: microphone unavailable, continuing without audio (${(err as Error).message})`);
  }

  console.log(`camera-record: capturing at ${video.width}x${video.height}`);

  // Defaults for H.264 at the capture's real geometry @30fps, then override bitrate.
  const config: AutoVideoEncodeConfig = {
    ...AutoVideoEncodeConfig.defaults("h264", video.width, video.height, { num: 1, den: FPS }),
    bitrateBps: 4_000_000,
  };

  let encoder;
  try {
    // Tries the best available backend on this machine (GPU Zero-Copy first,
    // CPU-upload fallback) — throws if none exists here.
    encoder = await openAutoEncoder(config);
  } catch (err) {
    if (err instanceof EncoderUnavailableError) {
      console.error(`camera-record: no encoder backend available on this platform (${err.message})`);
      video.close();
      audio?.close();
      return;
    }
    throw err;
  }

  const session = new EncodeSession(encoder);
  try {
    try {
      await record(video, audio, session, DURATION_MS);
    } finally {
      video.close();
      audio?.close();
    }

    // Flushes the encoder, finalizes the muxer, and returns the complete MP4 bytes.
    const mp4Bytes = await session.finish();

    await writeFile(OUTPUT_PATH, mp4Bytes);

    console.log(
      `camera-record: ${(DURATION_MS / 1000).toFixed(0)}s @ ${video.width}x${video.height} -> ${OUTPUT_PATH} (${mp4Bytes.length} bytes)`,
    );
  } finally {
    session.close();
  }
}

await main();
